pymdp/envs/rollout.py

```diff
@@ -34,134 +104,75 @@ def rollout(agent: Agent, env: Env, num_timesteps: int, rng_key: jr.PRNGKey, pol
 
     # default policy search just uses standard active inference policy selection
     if policy_search is None:
-
-        def default_policy_search(agent, qs, rng_key):
-            qpi, _ = agent.infer_policies(qs) # infer_policies computes posterior over policies using EFE
-            return qpi, None
-
         policy_search = default_policy_search
 
-    def step_fn(carry, x):
+    def step_fn(carry, t):
         # carrying the current timestep's action, observation, beliefs, empirical prior, environment state, and random key
-        action_t = carry["action_t"]
-        observation_t = carry["observation_t"]
+        action = carry["action"]
+        observation = carry["observation"]
         qs_prev = carry["qs"]
-        empirical_prior = carry["empirical_prior"]
         env = carry["env"]
         agent = carry["agent"]
         rng_key = carry["rng_key"]
 
-        # perform state inference using variational inference (FPI) - uses A matrix to map between hidden states and observations
-        qs = agent.infer_states(
-            observations=observation_t,
-            empirical_prior=empirical_prior,
-        )
+        keys = jr.split(rng_key, batch_size + 2)
+        rng_key = keys[0] # carry first key
 
-        rng_key, key = jr.split(rng_key)
-        qpi, _ = policy_search(agent, qs, key) # compute policy posterior using EFE - uses C to consider preferred outcomes
+        # infer next action and beliefs using the agent's inference and planning methods
+        # first timestep we don't have a prev action yet, so then we call with action=None
+        action_next, qs, xtra = jax.lax.cond(t > 0, lambda : infer_and_plan(agent, qs_prev, observation, action, keys[1], policy_search=policy_search), lambda : infer_and_plan(agent, qs_prev, observation, None, keys[1], policy_search=policy_search))
```

Comment:
perhaps we shouldn't check on `t` value, as one might want to provide a valid action with init_carry, and use this to calculate an empirical prior instead of agent.D ? :thinking: